File: mariadb-api/src/Controllers/ProjectFormsController.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Controllers;

use Enkl\Api\Db\Database;
use Enkl\Api\Services\FormService;
use Enkl\Api\Services\FormSubmissionService;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

/**
 * Ported from php-api/src/Controllers/ProjectFormsController.php (itself ported from Controllers/ProjectFormsController.cs). Project-member-facing surface for Enterprise
 * Forms — read published forms available to fill out, and manage the caller's own submission
 * drafts. No authoring/CRUD on Forms themselves lives here (mirrors ProjectStrategyController) —
 * every Form write happens through FormsController (OrgAdmin). Submit/approve actions land in
 * Phase 4/5; Phase 1 only has Draft CRUD.
 */
final class ProjectFormsController extends BaseController
{
    private function forms(): FormService
    {
        return new FormService(Database::connection());
    }

    private function submissions(): FormSubmissionService
    {
        return new FormSubmissionService(Database::connection());
    }

    public function listPublished(Request $request, Response $response): Response
    {
        return $this->json($response, $this->forms()->listPublished($this->callerOrgId($request)));
    }

    public function listMySubmissions(Request $request, Response $response, array $args): Response
    {
        return $this->json($response, $this->submissions()->listMine($args['projectId'], $this->callerUserId($request)));
    }

    public function listAwaitingMyAction(Request $request, Response $response, array $args): Response
    {
        return $this->json($response, $this->submissions()->listAwaitingMyAction($args['projectId'], $this->callerUserId($request), $this->callerIsOrgAdmin($request)));
    }

    public function getSubmission(Request $request, Response $response, array $args): Response
    {
        $result = $this->submissions()->get($args['projectId'], $args['submissionId']);
        return $result !== null ? $this->json($response, $result) : $this->notFound($response);
    }

    public function createSubmission(Request $request, Response $response, array $args): Response
    {
        $result = $this->submissions()->create($args['projectId'], $this->callerUserId($request), $this->body($request));
        return $result !== null ? $this->json($response, $result) : $this->notFound($response);
    }

    public function updateSubmission(Request $request, Response $response, array $args): Response
    {
        $result = $this->submissions()->update($args['projectId'], $this->callerUserId($request), $args['submissionId'], $this->body($request));
        return $result !== null ? $this->json($response, $result) : $this->notFound($response);
    }

    public function deleteSubmission(Request $request, Response $response, array $args): Response
    {
        $deleted = $this->submissions()->delete($args['projectId'], $this->callerUserId($request), $args['submissionId']);
        return $deleted ? $this->noContent($response) : $this->notFound($response);
    }

    public function submitSubmission(Request $request, Response $response, array $args): Response
    {
        $result = $this->submissions()->submit($args['projectId'], $this->callerUserId($request), $this->callerIsOrgAdmin($request), $args['submissionId']);
        if (!$result['ok']) {
            return $result['error'] === 'not_found' ? $this->notFound($response) : $this->json($response, ['message' => $result['error']], 400);
        }
        return $this->json($response, $result['dto']);
    }

    public function actOnApproval(Request $request, Response $response, array $args): Response
    {
        $body = $this->body($request);
        $result = $this->submissions()->actOnApproval(
            $args['projectId'], $this->callerUserId($request), $this->callerIsOrgAdmin($request),
            $args['submissionId'], (string) ($body['action'] ?? ''), $body['comment'] ?? null
        );
        if (!$result['ok']) {
            return $result['error'] === 'not_found' ? $this->notFound($response) : $this->json($response, ['message' => $result['error']], 400);
        }
        return $this->json($response, $result['dto']);
    }
}

File: api/Enkl.Api/Controllers/ProjectFormsController.cs
using Enkl.Api.Auth;
using Enkl.Api.Dtos;
using Enkl.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Enkl.Api.Controllers;

/// <summary>Project-member-facing surface for Enterprise Forms — read published forms available to
/// fill out, and manage the caller's own submission drafts. No authoring/CRUD on Forms themselves
/// lives here at all (mirrors ProjectStrategyController's own doc comment) — every Form write
/// happens through FormsController (OrgAdmin). Submit/approve actions land in Phase 4/5 once the
/// workflow engine exists to evaluate them; Phase 1 only has Draft CRUD.</summary>
[ApiController]
[Authorize(Policy = "ProjectMember")]
[Route("api/projects/{projectId:guid}/forms")]
public class ProjectFormsController : ControllerBase
{
    private readonly FormService _forms;
    private readonly FormSubmissionService _submissions;

    public ProjectFormsController(FormService forms, FormSubmissionService submissions)
    {
        _forms = forms;
        _submissions = submissions;
    }

    [HttpGet]
    public async Task<IActionResult> ListPublished()
    {
        return Ok(await _forms.ListPublishedAsync(User.OrgId()));
    }

    private bool CallerIsOrgAdmin => User.HasClaim("orgAdmin", "true");

    [HttpGet("submissions/mine")]
    public async Task<IActionResult> ListMySubmissions(Guid projectId)
    {
        return Ok(await _submissions.ListMineAsync(projectId, User.UserId()));
    }

    [HttpGet("submissions/awaiting-me")]
    public async Task<IActionResult> ListAwaitingMyAction(Guid projectId)
    {
        return Ok(await _submissions.ListAwaitingMyActionAsync(projectId, User.UserId(), CallerIsOrgAdmin));
    }

    [HttpGet("submissions/{submissionId:guid}")]
    public async Task<IActionResult> GetSubmission(Guid projectId, Guid submissionId)
    {
        var result = await _submissions.GetAsync(projectId, submissionId);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPost("submissions")]
    public async Task<IActionResult> CreateSubmission(Guid projectId, CreateFormSubmissionRequest request)
    {
        var result = await _submissions.CreateAsync(projectId, User.UserId(), request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpPut("submissions/{submissionId:guid}")]
    public async Task<IActionResult> UpdateSubmission(Guid projectId, Guid submissionId, UpdateFormSubmissionRequest request)
    {
        var result = await _submissions.UpdateAsync(projectId, User.UserId(), submissionId, request);
        return result is null ? NotFound() : Ok(result);
    }

    [HttpDelete("submissions/{submissionId:guid}")]
    public async Task<IActionResult> DeleteSubmission(Guid projectId, Guid submissionId)
    {
        return await _submissions.DeleteAsync(projectId, User.UserId(), submissionId) ? NoContent() : NotFound();
    }

    [HttpPost("submissions/{submissionId:guid}/submit")]
    public async Task<IActionResult> Submit(Guid projectId, Guid submissionId)
    {
        var (ok, error, dto) = await _submissions.SubmitAsync(projectId, User.UserId(), CallerIsOrgAdmin, submissionId);
        if (!ok) return error == "not_found" ? NotFound() : BadRequest(new { message = error });
        return Ok(dto);
    }

    [HttpPost("submissions/{submissionId:guid}/approval-action")]
    public async Task<IActionResult> ActOnApproval(Guid projectId, Guid submissionId, FormApprovalActionRequest request)
    {
        var (ok, error, dto) = await _submissions.ActOnApprovalAsync(projectId, User.UserId(), CallerIsOrgAdmin, submissionId, request.Action, request.Comment);
        if (!ok) return error == "not_found" ? NotFound() : BadRequest(new { message = error });
        return Ok(dto);
    }
}
